### DISCUSSION TITLE
convert2rhel through a proxy

### DISCUSSION THREAD

**Discussion by manica** [oamg/convert2rhel#261] [Ideas]
is it possible to add proxy parameter on convert2rhel command ?
like --http-proxy=AAA:port --https-prxy=BBB:port
thanks

**Comment by bocekm** [comment]
Hi Andrea, thanks for the question.

- In case you want convert2rhel to get to RHEL repos through RHSM with proxy set up, unfortunately there's currently no way to set that :(
  - From what I've read at https://access.redhat.com/solutions/57669, convert2rhel would need to run a command like `subscription-manager config --server.proxy_hostname=proxy.example.com --server.proxy_port=8080 --server.proxy_user=admin --server.proxy_password=secret`
- In case you use custom RHEL repositories for the conversion, it should be enough to add the information about the proxy server into /etc/yum.conf. Even though I haven't tried that.

Anyway, I'm adding the need to allow proxy with RHSM to our todo list ([OAMG-4785](https://issues.redhat.com/browse/OAMG-4785)).

**Comment by bocekm** [comment]
There's a workaround suggested by Loren under https://access.redhat.com/articles/2360841#comment-2038731 which I haven't tested though:

> After launching the tool, get to the point where subscription-manager is deployed. It should fail to register and re-prompt you. Alternatively, do what I did, and neglect to enter the password because I don't like having that on a command-line. Once you're at the subscription manger prompt for username/password, open another terminal, configure the proxy info in /etc/rhsm/rhsm.conf, and it'll continue.

**Reply by jbroome** [reply]
That doesn't really lend itself to automation though.

**Reply by bocekm** [reply]
@jbroome, see https://github.com/oamg/convert2rhel/discussions/261#discussioncomment-794649.

**Comment by bocekm** [comment]
Here's another workaround. This one is easier to automate.

1. Configure /etc/yum.conf to make yum use a proxy per https://access.redhat.com/solutions/4200391
2. Install subscription-manager
 - on CentOS Linux 7 and 8, subscription-manager is available in the standard repos, so just run `yum install subscription-manager`
 - on CentOS Linux 6, enable the contrib repo: `yum install --enablerepo contrib subscription-manager`
 - on Oracle Linux, add one of the following repositories first, based on your system major version
```
[centos-6-contrib-convert2rhel]
name=CentOS Linux 6 - Contrib
baseurl=https://vault.centos.org/centos/6/contrib/$basearch/
gpgcheck=0
enabled=1

[ubi-7-convert2rhel]
name=Red Hat Universal Base Image 7
baseurl=https://cdn-ubi.redhat.com/content/public/ubi/dist/ubi/server/7/7Server/$basearch/os/
gpgcheck=0
enabled=1

[ubi-8-baseos-convert2rhel]
name=Red Hat Universal Base Image 8 - BaseOS
baseurl=https://cdn-ubi.redhat.com/content/public/ubi/dist/ubi8/8/$basearch/baseos/os/
gpgcheck=0
enabled=1
```
3. Edit the proxy settings in /etc/rhsm/rhsm.conf per https://access.redhat.com/solutions/57669
4. Run `chattr +i /etc/rhsm/rhsm.conf` to make it undeletable (convert2rhel is removing any installed subscription-manager)
5. Run convert2rhel

**Reply by wp-perc** [reply]
I am a bit late, but... this is simply not serious. Are you expecting no enterprise class customer with a proxy to access RHEL repos? I hope there is another solution...

**Comment by bocekm** [comment]
Hi @wp-perc, customers have systems in all kinds of environments and we try to prioritize the cases based on how many customers are affected and how critical is it for them. "Enterprise class" customers mostly use an on-premise Red Hat Satellite 6 server to provide access to RHEL repositories and we rarely see proxy being used.

That said, last year [we added a new `--keep-rhsm` option](https://github.com/oamg/convert2rhel/pull/294) which helps with simplifying the proxy workaround a little bit by removing the need for setting the `i` attribute for the `rhsm.conf`:

1. Make sure you've configured `/etc/yum.conf` for yum to use a proxy per https://access.redhat.com/solutions/4200391
2. Configure one of the following repositories:
- on RHEL 7-based systems:
```
[ubi-for-convert2rhel]
name=Red Hat Universal Base Image 7
baseurl=https://cdn-ubi.redhat.com/content/public/ubi/dist/ubi/server/7/7Server/$basearch/os/
gpgcheck=0
enabled=1
```
- on RHEL 8-based systems:
```
[ubi-for-convert2rhel]
name=Red Hat Universal Base Image 8 - BaseOS
baseurl=https://cdn-ubi.redhat.com/content/public/ubi/dist/ubi8/8/$basearch/baseos/os/
gpgcheck=0
enabled=1
```
3. Install subscription-manager
```
# yum install -y --disablerepo="*" --enablerepo ubi-for-convert2rhel subscription-manager
```
4. Edit the proxy settings in `/etc/rhsm/rhsm.conf` per https://access.redhat.com/solutions/57669
5. Run convert2rhel with the `--keep-rhsm` option

**Comment by tuxmaster5000** [comment]
I fall into the same, but nothing will work.
If the system can direct access the internet it will work.
Behind the proxy it can't get the repositories for the activation_key.
Like without proxy I have added the org and activation_key to the /etc/convert2rhel.ini file.
The --keep-rhsm is unknown to convert2rhel.

**Reply by bocekm** [reply]
Hi @tuxmaster5000. We have posted workaround steps for conversions with a proxy in our docs https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/8/html-single/converting_from_an_rpm-based_linux_distribution_to_rhel/index#ref_known-issues-and-limitations_assembly_troubleshooting-rhel-conversions. If you haven't, please try those. The steps I described in this thread are outdated.

**Comment by tuxmaster5000** [comment]
Yes using the linked documentation will works. Thanks for the hint.
